[1.10.x] Added a way to create horse armor.

Darkhax opened the pull request:
**The issue being solved**
Horse armor is currently handled through hard references, which makes it impractical for mods to add horse armor. Past attempts usually involve ASM or replacing the horse entity all together, which are not very friendly when it comes to compatibility. 

**How it's being solved**
This PR adds the IHorseArmor interface, which can be implemented by items. This interface gives the item the ability to return a HorseArmorType enum, which can change the texture and protection amount offered. Items with this interface will also be accepted in the horse armor slot. HorseArmorType has been patched to take IHorseArmor into consideration. EntityHorse has also been patched to track the name of the HorseArmorType it has. This allows the related code to retrieve the HorseArmorType by name rather than ordinal. Name lookup is to allow mods to add entries without worrying about load order or being unloaded.

**Use Case**
- Mods will be able to add horse armor, to match the player armor they already have.
- Mods would be able to create custom calvery. Things like Thaumcraft's Crimson Knights riding horses with Crimson armor would be possible.
- Functional horse armor such as horse jet packs or horse wings would be possible. 

**Test Info**
A test mod is included with this PR. The test mod adds a basic piece of horse armor which can be found in the Misc Creative Tab. This armor should give a horse 23 armor points. You can use the following command to spawn a pre-tamed horse at your location. `/summon EntityHorse ~ ~ ~ {Tame:1,SaddleItem:{id:saddle,Count:1}}`

Darkhax committed:
Added a way to create horse armor.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=3185) <br/>All committers have signed the CLA.

Darkhax committed:
Fixed indents

Ellpeck commented:
This is a really awesome idea. I could implement this in one of my mods!

iLexiconn commented on patches/minecraft/net/minecraft/entity/passive/HorseArmorType.java.patch:
```diff
@@ -0,0 +1,22 @@
+ 
+     public static HorseArmorType func_188580_a(@Nullable ItemStack p_188580_0_)
+     {
++    	if (p_188580_0_ != null && p_188580_0_.func_77973_b() instanceof net.minecraftforge.common.IHorseArmor) return ((net.minecraftforge.common.IHorseArmor) p_188580_0_.func_77973_b()).getArmorType(p_188580_0_);
```
idk what character that is but it's not a space

iLexiconn commented on src/main/java/net/minecraftforge/common/IHorseArmor.java:
```diff
@@ -0,0 +1,37 @@
+ */
+public interface IHorseArmor 
+{
+
```
also use spaces here

iLexiconn commented on src/test/java/net/minecraftforge/test/HorseArmorTest.java:
```diff
@@ -0,0 +1,80 @@
+    }
+    public class ItemTestHorseArmor extends Item implements IHorseArmor 
+    {
+    	public ItemTestHorseArmor() {
```
what's up with the formatting in this class?

Darkhax committed:
Disabled test mod by default

heldplayer commented:
Does this keep compat with vanilla clients when connecting to a Forge server?

Darkhax commented:
@heldplayer After testing, it seems this PR does break that support. Looks like I will have to find some other way to do textures.

Darkhax committed:
Removed additional patches from EntityHorse, only edit the enum.

kashike commented on src/main/java/net/minecraftforge/common/IHorseArmor.java:
```diff
@@ -0,0 +1,20 @@
+package net.minecraftforge.common;
```
Missing license header.

Darkhax replied:
Added that in. Don't think GitHub will outdate this for me though.

Darkhax committed:
Added missing license header.

Darkhax commented:
Removed the patches to EntityHorse, it seems syncing the worn item will break vanilla support. As there is no way to sync the armor stack, mods will have to use the ItemStack from IHorseArmor#getArmorType to switch textures. I have also added a new constructor to HorseArmorType to allow mods to use a custom domain for their textures.

LexManos commented:
This is not really fasible to do this way as stated it breaks vanilla clients.
Horse with unknown armor == ArrayIndexOutOfBounds.
You can add extra data manager objects to sync to the client and vanilla will happily eat them without any concern.
Similar to how we enhance the Villagers/Zombies.

Darkhax committed:
Disabled the test by default again.

Darkhax committed:
Switched to string based lookup. Fixes a few issues.

- Fixed ArrayIndexOutOfBounds on missing armor type.
- Fixed mod load order allowing the type to change

Darkhax committed:
Prevent non-vanilla armors from being stored in int form.

Darkhax committed:
Fixed vanilla horse armor not syncing.

Darkhax commented:
That makes sense, didn't account for mods being installed, or load order changes originally. The latest commits use string based lookup, where the string is the name of the constant in `HorseArmorType`. Was thinking about creating a separate registry or lookup specifically for this, however that would likely result in more patching. It is also not necessary to read/write this name from NBT, as it is automatically applied when the armor stack is read from NBT. 

Edit: I have tested this version against load order changes with multiple mods adding horse armor, uninstalling a mod with a certain type of armor, and vanilla client connection. All seem to work as intended.

kashike commented on patches/minecraft/net/minecraft/entity/passive/EntityHorse.java.patch:
```diff
@@ -1,6 +1,40 @@
+     private static final DataParameter<Integer> field_184789_bG = EntityDataManager.<Integer>func_187226_a(EntityHorse.class, DataSerializers.field_187192_b);
+     private static final DataParameter<Optional<UUID>> field_184790_bH = EntityDataManager.<Optional<UUID>>func_187226_a(EntityHorse.class, DataSerializers.field_187203_m);
+     private static final DataParameter<Integer> field_184791_bI = EntityDataManager.<Integer>func_187226_a(EntityHorse.class, DataSerializers.field_187192_b);
++    private static final DataParameter<String> ARMOR_STRING = EntityDataManager.<String>func_187226_a(EntityHorse.class, DataSerializers.field_187194_d);
```
Will vanilla clients have an issue with this new entry?

williewillus replied:
as long as new DataParameter's are last (aka we do not interfere with the ordering of the original ones), they will be fine and ignored by vanilla clients

Forge adds new DataParameter's to Villagers (or was it zombies?) to support custom variants

Actuarius commented:
@williewillus added labels [Feature]

Darkhax commented:
Are there any further issues with this PR?

jriwanek commented:
Would be useful to me for sure.

p455w0rd commented:
This would be awesome to finally be able to do

Darkhax commented:
So, it's been about 3 months since this PR was opened. I know you're busy @LexManos but I would really appreciate knowing your stance on this.

jeffryfisher commented:
Enabling horse armor has been one of my targets for over two years. If there's work to be done (e.g. testing/debugging), then call on me (git@jeffryfisher.net). If there's a provisional Forge build (for mc 1.10.2) with this feature, then I'd be willing to update a mod ASAP. One of my mods already has a craftable, enchantable horse armor item with texture files; it just doesn't play nice with the slot on the horses.

Traverse-Joe commented:
Agreed It would Be Amazing

Lanse505 commented:
I +10 this post

BunnyCinnamon commented:
As I raise my arms /(°^°)/ This shall become a thing!

HenryLoenwind commented:
Good thing to have. And while you're at it, you can continue with the Shield and the Elytra. ;)

LexManos commented on patches/minecraft/net/minecraft/entity/passive/HorseArmorType.java.patch:
```diff
@@ -0,0 +1,36 @@
+     }
++    
++    // Forge    
++    private HorseArmorType(int armorStengthIn, String textureName)
```
There is nothing in here deprecating the old methods, or killing off the usages of the old bad methods.
The constructor also has no documentation. 
Also by preliminary glance at how this class is used, it looks like you need some mechanic of specifying the domain for the texture in EntityHorse.setHorseTexturePaths

LexManos commented on src/main/java/net/minecraftforge/common/IHorseArmor.java:
```diff
@@ -0,0 +1,39 @@
+ * When implemented by an item, it will be allowed to function as a piece of
+ * horse armor.
+ */
+public interface IHorseArmor 
```
not a fan of single use interfaces like this... it definitely shouldn't just be slapped in common. It ONLY ever gunna be applicable to items, why not just make it a thing on item?

liach replied:
This should be changed into capabilities possibly.

LexManos commented on src/main/java/net/minecraftforge/common/IHorseArmor.java:
```diff
@@ -0,0 +1,39 @@
+     * @param stack The ItemStack instance of the armor.
+     * @return HorseArmorType The HorseArmorType to use for this armor.
+     */
+    HorseArmorType getArmorType(ItemStack stack);
```
Needs a @Nullable as this is called in places where you only have the Item and not the stack, unless you go through and fix that.

LexManos commented on src/main/java/net/minecraftforge/common/util/EnumHelper.java:
```diff
@@ -123,6 +125,10 @@ public static EnumRarity addRarity(String name, TextFormatting color, String dis
     {
         return addEnum(EnumRarity.class, name, color, displayName);
     }
+    public static HorseArmorType addArmorType(String name, int protection, String texture)
```
no docs no nothing esp on a constructor added by Forge.

LexManos commented:
Looks like you've missed A LOT of the places where armor type is used to determine things.
You're lacking documentation, and the single purpose interfaces are not an acceptible design for Forge.
Its fine in other things but we've gone the route of just putting them as methods instead of constantly having to do InstanceOf

mezz commented:
Please update this to address Lex's comments and to target 1.11, new features are no longer being merged into 1.10.

mezz commented:
Any update @darkhax ?

mezz commented:
I am closing this since you have not responded in a long time.
If you decide to update this please open a new PR and link back to this one.
